tls-client/main.cpp

```diff
@@ -196,108 +216,118 @@ class HelloHTTPS {
                 print_mbedtls_error("mbedtls_ssl_write", ret);
-                onError(_tcpsocket, -1 );
+                onError(_tcpsocket, ret);
             }
-            return;
+            goto close_tls;
         }
 
         /* It also means the handshake is done, time to print info */
         printf("TLS connection to %s established\r\n", HTTPS_SERVER_NAME);
 
-        const uint32_t buf_size = 1024;
-        char *buf = new char[buf_size];
-        mbedtls_x509_crt_info(buf, buf_size, "\r    ",
+        mbedtls_x509_crt_info(_crt_buf, sizeof(_crt_buf) - 1, "\r    ",
                         mbedtls_ssl_get_peer_cert(&_ssl));
-        mbedtls_printf("Server certificate:\r\n%s\r", buf);
+        mbedtls_printf("Server certificate:\r\n%s\r", _crt_buf);
 
-        uint32_t flags = mbedtls_ssl_get_verify_result(&_ssl);
-        if( flags != 0 )
-        {
-            mbedtls_x509_crt_verify_info(buf, buf_size, "\r  ! ", flags);
-            printf("Certificate verification failed:\r\n%s\r\r\n", buf);
-        }
-        else
+        flags = mbedtls_ssl_get_verify_result(&_ssl);
+        if(flags != 0) {
+            mbedtls_x509_crt_verify_info(_crt_buf, sizeof(_crt_buf) - 1, "\r  ! ", flags);
+            printf("Certificate verification failed:\r\n%s\r\r\n", _crt_buf);
+        } else {
             printf("Certificate verification passed\r\n\r\n");
+        }
 
 
         /* Read data out of the socket */
-        ret = mbedtls_ssl_read(&_ssl, (unsigned char *) _buffer, sizeof(_buffer));
+        ret = mbedtls_ssl_read(&_ssl, (unsigned char *) _recv_buf, sizeof(_recv_buf));
         if (ret < 0) {
             if (ret != MBEDTLS_ERR_SSL_WANT_READ && ret != MBEDTLS_ERR_SSL_WANT_WRITE) {
                 print_mbedtls_error("mbedtls_ssl_read", ret);
-                onError(_tcpsocket, -1 );
+                onError(_tcpsocket, ret);
             }
-            delete[] buf;
-            return;
+            goto close_tls;
         }
         _bpos = static_cast<size_t>(ret);
 
-        _buffer[_bpos] = 0;
+        _recv_buf[_bpos] = 0;
 
         /* Check each of the flags */
-        _got200 = _got200 || strstr(_buffer, HTTPS_OK_STR) != NULL;
-        _gothello = _gothello || strstr(_buffer, HTTPS_HELLO_STR) != NULL;
+        _got200 = _got200 || strstr(_recv_buf, HTTPS_OK_STR) != NULL;
+        _gothello = _gothello || strstr(_recv_buf, HTTPS_HELLO_STR) != NULL;
 
         /* Print status messages */
         mbedtls_printf("HTTPS: Received %d chars from server\r\n", _bpos);
         mbedtls_printf("HTTPS: Received 200 OK status ... %s\r\n", _got200 ? "[OK]" : "[FAIL]");
         mbedtls_printf("HTTPS: Received '%s' status ... %s\r\n", HTTPS_HELLO_STR, _gothello ? "[OK]" : "[FAIL]");
         mbedtls_printf("HTTPS: Received message:\r\n\r\n");
-        mbedtls_printf("%s", _buffer);
+        mbedtls_printf("%s", _recv_buf);
         _error = !(_got200 && _gothello);
```

Comment:
_error should be set at all places where we jump to exit. Else it will always print DONE.